### ISSUE TITLE
exported name not exposed from workspace package even though listed in __all__

### ISSUE THREAD

**Issue by smackesey** [microsoft/pylance-release#2150]
I am working with a Python library [Dagster](https://github.com/dagster-io/dagster). The recommended way to use Dagster is to import from the top-level module, like this:

```python
from dagster import job
```

`job`(and Dagster's other constructs) are exposed in the top level dagster module by importing them from submodules and listing them in `__all__`. Here is what the chain of imports looks like that ultimately ends up exposing `job`:

```python
##### dagster/__init__.py

from dagster.core.definitions import job

__all__ = [
    'job',
]


##### dagster/core/definitions/__init__.py

from .decorators import job


##### dagster/core/definitions/decorators/__init__.py

from .job import job


##### dagster/core/definitions/decorators/job.py

def job():
    # ... code ...
```

When I am coding and try to auto-complete `job`, the only candidate I see is from the place where `job` is actually defined in the code: `dagster/core/definitions/decorators/job`. I would expect to be able to import from `dagster`, since `job` is listed in the `__all__`. Is this a bug?

Note that the `dagster` package is in my workspace via `extraPaths`, and I am running `pyright` 1.1.193.

**Comment by smackesey** [comment]
I've noticed that this bug does not manifest in `pylance` (not sure if it ever did), but the buggy behavior is still present in `pyright`.

**Comment by erictraut** [comment]
I think this is an expected behavior in pyright. The auto-import suggestions are limited to those symbols that pyright has encountered naturally as part of its type evaluations, which are done lazily. It doesn't explore libraries to discover additional potentially-importable symbols. Pylance adds this functionality by way of a background indexing mechanism. We currently have no plans to backport this mechanism to pyright. If you want this functionality, we recommend using pylance.

I think this bug can be closed since it's not a bug in pylance, and it's expected behavior in pyright.

**Comment by smackesey** [comment]
@erictraut In the other thread where we discussed this, you said:

>If the same name is exported from multiple modules, the auto-import logic in pyright and pylance prefer the shortest module path. Only the shortest path is listed in the completion list, and the longer paths are de-duped. Currently there's no way for a library author to override that heuristic. So far, that heuristic seems to work well.

What you're saying here seems inconsistent with this. `pyright` _consistently_ auto-imports from the _longest_ path even when it has already scanned the shorter path in which the symbol is defined.

**Comment by erictraut** [comment]
Yes, you're correct. I'm able to repro the behavior you're describing, and it's not what I would expect.

@heejaechang, the problem is due to the following check in `autoImporter.ts`:
```python
                        if (declaration.type === DeclarationType.Alias) {
                            // We don't include import alias in auto import
                            // for workspace files.
                            return;
                        }
```

This comment indicates that this logic should apply only to user code, not to libraries, but it is applied in all cases.

The reason this works in pylance is that it uses the separate library index (if indexing is enabled). The same bug would probably manifest in pylance if indexing is disabled.

**Comment by erictraut** [comment]
I've confirmed that when I comment out the above lines, I observe the expected behavior.

**Comment by smackesey** [comment]
This is still an issue with `pyright`. We recently started including `py.typed` in Dagster, and now this bug is causing pain for our users. Do you expect this to be resolved anytime soon? Thanks.

**Comment by erictraut** [comment]
@heejaechang, do you need help with this? I would think this is a straightforward bug to fix within the auto import logic. See above for my diagnosis and recommendation.

**Comment by debonte** [comment]
@smackesey, I believe @heejaechang started looking at this. He's out this week though. Expect an update from him next week.

**Comment by smackesey** [comment]
@heejaechang Just wondering if you've started looking at this-- thanks

**Comment by smackesey** [comment]
Sorry to be annoying, but I've got to keep pinging about this-- this is leading our users to autoimport from non-public submodules.

**Comment by debonte** [comment]
I'll take a look at this today.

**Comment by debonte** [comment]
@smackesey, the fix will be in the next release of both Pyright and Pylance.
